### PR TITLE
Add metrics and dynamic config constants for callbacks

### PR DESCRIPTION
**Pull request by bergundy**
Prep work for Nexus callbacks.

### CODE REVIEW

File: common/dynamicconfig/constants.go [hunk 1]
```diff
@@ -673,6 +673,41 @@ const (
...
+	// CallbackProcessorFailoverMaxPollRPS is max poll rate per second for callbackQueueFactory
+	CallbackProcessorFailoverMaxPollRPS = "history.callbackProcessorFailoverMaxPollRPS"
+	// CallbackProcessorMaxPollRPS is max poll rate per second for callbackQueueFactory
+	CallbackProcessorMaxPollRPS = "history.callbackProcessorMaxPollRPS"
+	// CallbackProcessorMaxPollHostRPS is max poll rate per second for all callbackQueueFactory on a host
+	CallbackProcessorMaxPollHostRPS = "history.callbackProcessorMaxPollHostRPS"
+	// CallbackProcessorSchedulerWorkerCount is the number of workers in the host level task scheduler for callbackQueueFactory
+	CallbackProcessorSchedulerWorkerCount = "history.callbackProcessorSchedulerWorkerCount"
+	// CallbackProcessorSchedulerActiveRoundRobinWeights is the priority round robin weights used by callback task scheduler for active namespaces
+	CallbackProcessorSchedulerActiveRoundRobinWeights = "history.callbackProcessorSchedulerActiveRoundRobinWeights"
```
Comment (ast2023) [inline common/dynamicconfig/constants.go:687]: I think a more consistent name would be either `CallbackProcessorActiveSchedulerRoundRobinWeights` or `CallbackProcessorSchedulerRoundRobinWeightsActive`. I prefer `ActiveScheduler`. The reason is that there is no `CallbackProcessorSccheduler` setting or struct here. So, I would go with proper English.

File: common/dynamicconfig/constants.go [hunk 2]
```diff
@@ -673,6 +673,41 @@ const (
...
+	// CallbackProcessorSchedulerStandbyRoundRobinWeights is the priority round robin weights used by callback task scheduler for standby namespaces
+	CallbackProcessorSchedulerStandbyRoundRobinWeights = "history.callbackProcessorSchedulerStandbyRoundRobinWeights"
+	// CallbackProcessorUpdateShardTaskCount is update shard count for callbackQueueFactory
+	CallbackProcessorUpdateShardTaskCount = "history.callbackProcessorUpdateShardTaskCount"
```
Comment (ast2023) [inline common/dynamicconfig/constants.go:691]: `UpdateShardTaskCount` is a verb group, a better name would be `CountForUpdateShardTask`

**Review by ast2023 — APPROVED** [review]

File: common/dynamicconfig/constants.go [hunk 3]
```diff
@@ -673,6 +673,41 @@ const (
...
+	// CallbackProcessorMaxPollInterval max poll interval for callbackQueueFactory
+	CallbackProcessorMaxPollInterval = "history.callbackProcessorMaxPollInterval"
+	// CallbackProcessorMaxPollIntervalJitterCoefficient is the max poll interval jitter coefficient
+	CallbackProcessorMaxPollIntervalJitterCoefficient = "history.callbackProcessorMaxPollIntervalJitterCoefficient"
+	// CallbackProcessorUpdateAckInterval is update interval for callbackQueueFactory
+	CallbackProcessorUpdateAckInterval = "history.callbackProcessorUpdateAckInterval"
```
[the remark this answers is not in the document]
Comment (pdoerner) [reply common/dynamicconfig/constants.go:697]: Personally, I prefer the consistency with our other similarly named options

**Review by pdoerner — APPROVED** [review]

File: common/dynamicconfig/constants.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ast2023) [reply common/dynamicconfig/constants.go:697]: @pdoerner not sure if you wanted to put your comment here, but the name UpdateAckInterval does not follow the convention all other fields follow. Every field name matches more or less what the comment says. This one is different.

File: service/history/configs/config.go [hunk 4]
```diff
@@ -444,6 +462,23 @@ func NewConfig(
 		TransferProcessorEnsureCloseBeforeDelete:            dc.GetBoolProperty(dynamicconfig.TransferProcessorEnsureCloseBeforeDelete, true),
 		TimerQueueMaxReaderCount:                            dc.GetIntProperty(dynamicconfig.TimerQueueMaxReaderCount, 2),
 
+		CallbackTaskBatchSize:                               dc.GetIntProperty(dynamicconfig.CallbackTaskBatchSize, 100),
+		CallbackProcessorSchedulerWorkerCount:               dc.GetIntProperty(dynamicconfig.CallbackProcessorSchedulerWorkerCount, 512),
+		CallbackProcessorSchedulerActiveRoundRobinWeights:   dc.GetMapPropertyFnWithNamespaceFilter(dynamicconfig.CallbackProcessorSchedulerActiveRoundRobinWeights, ConvertWeightsToDynamicConfigValue(DefaultActiveTaskPriorityWeight)),
+		CallbackProcessorSchedulerStandbyRoundRobinWeights:  dc.GetMapPropertyFnWithNamespaceFilter(dynamicconfig.CallbackProcessorSchedulerStandbyRoundRobinWeights, ConvertWeightsToDynamicConfigValue(DefaultStandbyTaskPriorityWeight)),
+		CallbackProcessorFailoverMaxPollRPS:                 dc.GetIntProperty(dynamicconfig.CallbackProcessorFailoverMaxPollRPS, 1),
```
Comment (yycptt) [inline service/history/configs/config.go:469]: FailoverMaxPollRPS is no longer needed.

File: service/history/configs/config.go [hunk 5]
```diff
@@ -444,6 +462,23 @@ func NewConfig(
...
+		CallbackProcessorMaxPollHostRPS:                     dc.GetIntProperty(dynamicconfig.CallbackProcessorMaxPollHostRPS, 0),
+		CallbackProcessorMaxPollInterval:                    dc.GetDurationProperty(dynamicconfig.CallbackProcessorMaxPollInterval, 1*time.Minute),
+		CallbackProcessorMaxPollIntervalJitterCoefficient:   dc.GetFloat64Property(dynamicconfig.CallbackProcessorMaxPollIntervalJitterCoefficient, 0.15),
+		CallbackProcessorUpdateAckInterval:                  dc.GetDurationProperty(dynamicconfig.CallbackProcessorUpdateAckInterval, 30*time.Second),
+		CallbackProcessorUpdateAckIntervalJitterCoefficient: dc.GetFloat64Property(dynamicconfig.CallbackProcessorUpdateAckIntervalJitterCoefficient, 0.15),
+		CallbackProcessorCompleteCallbackInterval:           dc.GetDurationProperty(dynamicconfig.CallbackProcessorCompleteCallbackInterval, 60*time.Second),
+		CallbackProcessorPollBackoffInterval:                dc.GetDurationProperty(dynamicconfig.CallbackProcessorPollBackoffInterval, 5*time.Second),
+		CallbackProcessorEnsureCloseBeforeDelete:            dc.GetBoolProperty(dynamicconfig.CallbackProcessorEnsureCloseBeforeDelete, true),
```
Comment (yycptt) [inline service/history/configs/config.go:478]: I don't think callback processor need this. It's from tasks that need to delete something.

File: service/history/configs/config.go [hunk 6]
```diff
@@ -444,6 +462,23 @@ func NewConfig(
...
+		CallbackProcessorFailoverMaxPollRPS:                 dc.GetIntProperty(dynamicconfig.CallbackProcessorFailoverMaxPollRPS, 1),
+		CallbackProcessorMaxPollRPS:                         dc.GetIntProperty(dynamicconfig.CallbackProcessorMaxPollRPS, 20),
+		CallbackProcessorMaxPollHostRPS:                     dc.GetIntProperty(dynamicconfig.CallbackProcessorMaxPollHostRPS, 0),
+		CallbackProcessorMaxPollInterval:                    dc.GetDurationProperty(dynamicconfig.CallbackProcessorMaxPollInterval, 1*time.Minute),
+		CallbackProcessorMaxPollIntervalJitterCoefficient:   dc.GetFloat64Property(dynamicconfig.CallbackProcessorMaxPollIntervalJitterCoefficient, 0.15),
+		CallbackProcessorUpdateAckInterval:                  dc.GetDurationProperty(dynamicconfig.CallbackProcessorUpdateAckInterval, 30*time.Second),
+		CallbackProcessorUpdateAckIntervalJitterCoefficient: dc.GetFloat64Property(dynamicconfig.CallbackProcessorUpdateAckIntervalJitterCoefficient, 0.15),
+		CallbackProcessorCompleteCallbackInterval:           dc.GetDurationProperty(dynamicconfig.CallbackProcessorCompleteCallbackInterval, 60*time.Second),
```
Comment (yycptt) [inline service/history/configs/config.go:476]: CompleteXXXInterval is also no longer needed.

File: common/metrics/metric_defs.go [hunk 7]
```diff
@@ -465,6 +474,8 @@ const (
 	OperationArchivalQueueProcessorScope = "ArchivalQueueProcessor"
 	// OperationMemoryScheduledQueueProcessorScope is a scope for memory scheduled queue processor.
 	OperationMemoryScheduledQueueProcessorScope = "MemoryScheduledQueueProcessor"
+	// OperationCallbackQueueProcessorScope is a scope for the callback queue processor.
+	OperationCallbackQueueProcessorScope = "CallbackQueueProcessor"
```
Comment (yycptt) [inline common/metrics/metric_defs.go:478]: Looks like we have some duplicated definition here...

File: common/dynamicconfig/constants.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply common/dynamicconfig/constants.go:697]: It should be `callbackQueueProcessor` not `callbackQueueFactory` in those comments I think.

File: common/metrics/metric_defs.go (hunk 7 above)
Comment (bergundy) [reply common/metrics/metric_defs.go:478]: Yup, same for transfer tasks and timer tasks BTW.

I just duplicated what we had for transfer tasks.

File: service/history/configs/config.go (hunk 4 above)
Comment (bergundy) [reply service/history/configs/config.go:469]: Thanks, I'll remove.

File: service/history/configs/config.go (hunk 6 above)
Comment (bergundy) [reply service/history/configs/config.go:476]: Thanks, I'll remove.

File: service/history/configs/config.go (hunk 5 above)
Comment (bergundy) [reply service/history/configs/config.go:478]: Thanks, I'll remove.

File: common/dynamicconfig/constants.go (hunk 1 above)
Comment (bergundy) [reply common/dynamicconfig/constants.go:687]: I agree we could use a review for the dynamic config structure and naming convention.
I'm just going with the same convention we have for the other processors.
If we want to change the naming, we should change all names and that should probably be in a PR that's not part of the feature branch.

File: common/dynamicconfig/constants.go (hunk 2 above)
Comment (bergundy) [reply common/dynamicconfig/constants.go:691]: Thanks, I'll remove.

File: common/dynamicconfig/constants.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply common/dynamicconfig/constants.go:697]: > It should be `callbackQueueProcessor` not `callbackQueueFactory` in those comments I think.

We have the processor terminology in other comments but nothing called processor in the code base.
Is this more of a theoretical concept?

**Review by bergundy — COMMENTED** [review]
Thanks for the comments everyone.

File: common/metrics/metric_defs.go (hunk 7 above)
Comment (yycptt) [reply common/metrics/metric_defs.go:478]: yeah.. I will clean them up.

**Pull request merged**